File: part2/exceptions/excep_client.cpp
// Apache Thrift User Defined Exception Client in C++

#include "gen-cpp/TradeHistory.h"
#include "gen-cpp/excep_types.h"
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TBufferTransports.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <boost/shared_ptr.hpp>
#include <boost/make_shared.hpp>
#include <iostream>

using namespace apache::thrift::transport;
using namespace apache::thrift::protocol;
using boost::shared_ptr;
using boost::make_shared;

int main(int argv, char * argc[]) {
    shared_ptr<TTransport> trans;
    trans = make_shared<TSocket>("localhost", 9090);
    trans = make_shared<TBufferedTransport>(trans);
    auto proto = make_shared<TBinaryProtocol>(trans);
    TradeHistoryClient client(proto);

    try {
        trans->open();
        auto price = client.GetLastSale(argc[1]);
        std::cout << "[Client] received: " << price << std::endl;
    } catch (const BadFish & bf) {
        std::cout << "[Client] GetLastSale() call failed for fish: "
                  << bf.fish << ", error: " << bf.error_code << std::endl;
    } catch (...) {
        std::cout << "[Client] GetLastSale() call failed" << std::endl;
    }
    trans->close();
}

File: part2/exceptions/ExcepClient.java
// Apache Thrift User Defined Exception Client in Java

import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.TException;

public class ExcepClient {
    public static void main(String[] args) throws TException {
        TSocket trans = new TSocket("localhost", 9090);
        TBinaryProtocol proto = new TBinaryProtocol(trans);
        TradeHistory.Client client = new TradeHistory.Client(proto);

        try {
            trans.open();
            double price = client.GetLastSale(args[0]);
            System.out.println("[Client] received: " + price);
        } catch (BadFish bf) {
            System.out.println("[Client] GetLastSale() call failed for fish: " + 
                               bf.fish + ", error " + bf.error_code);
        }
        trans.close();
    }
}

